src/core/or/circuitpadding.h

```diff
@@ -0,0 +1,539 @@
+  circpad_hist_bin_t histogram[CIRCPAD_MAX_HISTOGRAM_LEN];
+  /** total number of tokens */
+  uint32_t histogram_total;
+
+  /** Microseconds of the first bin of histogram, or base of iat dist */
+  circpad_delay_t start_usec;
+  /** The span of the histogram in seconds, used to calculate bin with.
+   *  For iat dist use, this is used as a max delay cap on the distribution. */
+  uint16_t range_sec;
+
+  /**
+   * The iat_dist is a parametrized way of encoding inter-packet delay
+   * information in microseconds. It can be used instead of histograms.
+   *
+   * If it is used, token_removal below must be set to
+   * CIRCPAD_TOKEN_REMOVAL_NONE.
+   *
+   * Start_usec, range_sec, and rtt_estimates are still applied to the
+   * results of sampling from this distribution (range_sec is used as a max).
+   */
+  circpad_distribution_t iat_dist;
+
+  /**
+   * The length dist is a parameterized way of encoding how long this
+   * state machine runs in terms of sent padding cells or all
+   * sent cells. Values are sampled from this distribution, clamped
+   * to max_len, and then start_len is added to that value.
+   *
+   * It may be specified instead of or in addition to
+   * the infinity bins and bins empty conditions. */
+  circpad_distribution_t length_dist;
+
+  /** A minimum length value, added to the output of length_dist */
+  uint16_t start_length;
+
+  /** A cap on the length value that can be sampled from the length_dist */
+  uint64_t max_length;
+
+  /** Should we decrement length when we see a nonpadding packet?
+   * XXX: Are there any machines that actually want to set this to 0? There may
+   * not be. OTOH, it's only a bit.. */
+  uint8_t length_includes_nonpadding : 1;
+
+  /**
+   * This is a bitfield that specifies which direction and types
+   * of traffic that cause us to remain in the current state. Cancel the
+   * pending padding packet (if any), and then await the next event.
+   *
+   * Example: Cancel padding if I saw a regular data packet.
+   */
+  circpad_event_t transition_cancel_events;
+
+  /**
+   * This is an array of bitfields that specifies which direction and
+   * types of traffic that cause us to abort our scheduled packet and
+   * switch to the state corresponding to the index of the array.
+   *
+   * Example: If the bins are empty (CIRCPAD_TRANSITION_ON_BINS_EMPTY) and that
+   * bit is set in the burst state index, then transition to the burst state.
+   */
+  circpad_event_t transition_events[CIRCPAD_NUM_STATES];
+
+  /**
+   * If true, estimate the RTT from this relay to the exit/website and add that
+   * to start_usec for use as the histogram bin 0 start delay.
+   *
+   * Right now this is only supported for relay-side state machines.
+   */
+  uint8_t use_rtt_estimate : 1;
+
+  /** This specifies the token removal strategy to use upon padding and
+   *  non-padding activity. */
+  circpad_removal_t token_removal;
+} circpad_state_t;
+
+/**
+ * Mutable padding machine info.
```

Comment:
"info" implies that we haven't thought this through completely.  Do we mean "state"?  Should the state type vary from machine to machine?

Reply:
The mutable piece is the important modifier here. This is all the information that we need to be able to update as part of running the state machine. Calling it the mutable "state" would be misleading -- some of the information here is associated with the current state, but some of it is associated with an entire padding state machine.

This whole struct is an optimization to reduce the need to re-allocate things that don't need to be re-allocated for each circuit (ie stuff in circpad_machine_t), and to make a strong barrier to prevent bugs from accidentally modifying global padding machines.